cats/cats-redis/src/test/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentSchedulerTest.java

```diff
@@ -0,0 +1,137 @@
+    DefaultAgentIntervalProvider intervalProvider = new DefaultAgentIntervalProvider(6000000);
+    clusteredSortAgentScheduler =
+        new ClusteredSortAgentScheduler(jedisPool, () -> false, intervalProvider, parallelism);
+
+    WAITING_SET =
+        (String)
+            FieldUtils.getDeclaredField(ClusteredSortAgentScheduler.class, "WAITING_SET", true)
+                .get(clusteredSortAgentScheduler);
+    WORKING_SET =
+        (String)
+            FieldUtils.getDeclaredField(ClusteredSortAgentScheduler.class, "WORKING_SET", true)
+                .get(clusteredSortAgentScheduler);
+    saturatePoolMethod =
+        MethodUtils.getMatchingMethod(ClusteredSortAgentScheduler.class, "saturatePool");
+    saturatePoolMethod.setAccessible(true);
+    runningAgents =
+        (Optional<Semaphore>)
+            FieldUtils.getDeclaredField(ClusteredSortAgentScheduler.class, "runningAgents", true)
+                .get(clusteredSortAgentScheduler);
+  }
+
+  @Test
+  public void testRunningAgentsSemaphore()
+      throws InvocationTargetException, IllegalAccessException {
+    when(jedis.zrangeByScore(eq(WORKING_SET), anyString(), anyString()))
+        .thenReturn(new HashSet<>());
+    when(jedis.zrangeByScore(eq(WAITING_SET), anyString(), anyString()))
+        .thenReturn(Set.of("testAgentType"));
+
+    saturatePoolMethod.invoke(clusteredSortAgentScheduler);
+
+    assertEquals(runningAgents.map(Semaphore::availablePermits).orElse(-1), parallelism);
+  }
+
+  @Test
+  public void testRunningAgentsSemaphoreWithException()
+      throws InterruptedException, InvocationTargetException, IllegalAccessException {
+    TestAgent agent1 = new TestAgent();
+    TestAgent agent2 = new TestAgent();
+    CountDownLatch latch = new CountDownLatch(1);
+
+    when(jedis.zrangeByScore(eq(WORKING_SET), anyString(), anyString()))
+        .thenReturn(new HashSet<>());
+    when(jedis.zrangeByScore(eq(WAITING_SET), anyString(), anyString()))
+        .thenReturn(Set.of(agent1.getAgentType(), agent2.getAgentType()));
+    when(jedis.scriptExists(anyString())).thenReturn(true);
+    clusteredSortAgentScheduler.schedule(agent1, agentExecution, executionInstrumentation);
+    clusteredSortAgentScheduler.schedule(agent2, agentExecution, executionInstrumentation);
+    when(jedis.evalsha(anyString(), anyList(), anyList()))
+        .thenReturn("testReleaseScore")
+        .thenThrow(new RuntimeException("fail"))
+        .thenReturn("testReleaseScore");
+    when(agentExecution.executeAgentWithoutStore(any()))
+        .thenReturn(new DefaultCacheResult(new HashMap<>()));
+    doAnswer(
+            (invocation) -> {
+              latch.countDown();
+              return null;
+            })
+        .when(agentExecution)
+        .storeAgentResult(any(), any());
+
+    saturatePoolMethod.invoke(clusteredSortAgentScheduler);
+    latch.await(10, TimeUnit.SECONDS);
+
+    assertEquals(runningAgents.map(Semaphore::availablePermits).orElse(-1), parallelism);
```

Comment:
I could use a comment here, specifically about the `orElse`.  Under what conditions does the orElse actually happen?

Reply:
In this [code](https://github.com/spinnaker/clouddriver/blob/909b6ff11dad82268a0b4401f70cabc9b4a3f3ed/cats/cats-redis/src/main/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentScheduler.java#L109), when parallelism is -1, the runningAgents is Optional.empty().

Reply:
No need.